examples/jsm/objects/BatchedMesh.js

```diff
@@ -0,0 +1,461 @@
+	DataTexture,
+	FloatType,
+	MathUtils,
+	Matrix4,
+	Mesh,
+	RGBAFormat
+} from 'three';
+
+const _identityMatrix = new Matrix4();
+const _zeroMatrix = new Matrix4().set(
+	0, 0, 0, 0,
+	0, 0, 0, 0,
+	0, 0, 0, 0,
+	0, 0, 0, 0
+);
+
+// Arbitary numbers
+const DEFAULT_MAX_GEOMETRY_COUNT = 256;
+const DEFAULT_MAX_VERTEX_COUNT = 1024 * 256;
+const DEFAULT_MAX_INDEX_COUNT = 1024 * 256;
+
+// Custom shaders
+const batchingParsVertex = `
+#ifdef BATCHING
+	attribute float id;
+	uniform highp sampler2D batchingTexture;
+	uniform int batchingTextureSize;
+	mat4 getBatchingMatrix( const in float i ) {
+		float j = i * 4.0;
+		float x = mod( j, float( batchingTextureSize ) );
+		float y = floor( j / float( batchingTextureSize ) );
+		float dx = 1.0 / float( batchingTextureSize );
+		float dy = 1.0 / float( batchingTextureSize );
+		y = dy * ( y + 0.5 );
+		vec4 v1 = texture2D( batchingTexture, vec2( dx * ( x + 0.5 ), y ) );
+		vec4 v2 = texture2D( batchingTexture, vec2( dx * ( x + 1.5 ), y ) );
+		vec4 v3 = texture2D( batchingTexture, vec2( dx * ( x + 2.5 ), y ) );
+		vec4 v4 = texture2D( batchingTexture, vec2( dx * ( x + 3.5 ), y ) );
+		mat4 bone = mat4( v1, v2, v3, v4 );
+		return bone;
+	}
+#endif
+`;
+
+const batchingbaseVertex = `
+#ifdef BATCHING
+	mat4 batchingMatrix = getBatchingMatrix( id );
+#endif
+`;
+
+const batchingnormalVertex = `
+#ifdef BATCHING
+	objectNormal = vec4( batchingMatrix * vec4( objectNormal, 0.0 ) ).xyz;
+	#ifdef USE_TANGENT
+		objectTangent = vec4( batchingMatrix * vec4( objectTangent, 0.0 ) ).xyz;
+	#endif
+#endif
+`;
+
+const batchingVertex = `
+#ifdef BATCHING
+	transformed = ( batchingMatrix * vec4( transformed, 1.0 ) ).xyz;
+#endif
+`;
+
+// @TODO: SkinnedMesh support?
+// @TODO: Move into the core. Can be optimized more with WEBGL_multi_draw.
+
+class BatchedMesh extends Mesh {
+
+	constructor( material, maxGeometryCount = DEFAULT_MAX_GEOMETRY_COUNT,
+		maxVertexCount = DEFAULT_MAX_VERTEX_COUNT, maxIndexCount = DEFAULT_MAX_INDEX_COUNT ) {
+
+		super( new BufferGeometry(), material );
+
+		this._vertexStarts = [];
+		this._vertexCounts = [];
+		this._indexStarts = [];
+		this._indexCounts = [];
+
+		this._visibilities = [];
+		this._alives = [];
+
+		this._maxGeometryCount = maxGeometryCount;
+		this._maxVertexCount = maxVertexCount;
+		this._maxIndexCount = maxIndexCount;
+
+		this._geometryInitialized = false;
+		this._currentGeometryCount = 0;
+		this._currentVertexCount = 0;
+		this._currentIndexCount = 0;
+
+		// Local matrix per geometry by using data texture
+		// @TODO: Support uniform parameter per geometry
+
+		this._matrices = []; 
+		this._matricesArray = null;
+		this._matricesTexture = null;
+		this._matricesTextureSize = null;
+
+		// @TODO: Calculate the entire binding box and make frustomCulled true
+		this.frustomCulled = false;
+
+		this._customUniforms = {
+			batchingTexture: { value: null },
+			batchingTextureSize: { value: 0 }
+		};
+
+		this._initMatricesTexture();
+		this._initShader();
+
+	}
```

Comment:
just an idea — it might be nice to have some getters...

- `.geometryCount`
- `.vertexCount`
- `.indexCount`

... so users can see how much space is left in the batch. The second two values would presumably change after removing a geometry and then calling `.optimize()`.

Reply:
Sounds good to me. Added the getters.

And I renamed `._currentGeometry/Vertex/IndexCount` to `._geometry/vertex/indexCount`. I don't really remember why I named them `_current`.